nexusclient/client.go

```diff
@@ -0,0 +1,725 @@
+func (h *OperationHandle) Operation() string {
+	return h.operation
+}
+
+// ID is the handler generated ID for this handle's operation.
+// Empty for synchronous operations.
+func (h *OperationHandle) ID() string {
+	return h.id
+}
+
+// State is the last known operation state.
+// Empty for handles created with [nexusclient.Client.GetHandle] before issuing a request to get the result.
+//
+// ⚠️ [nexusclient.OperationHandle.GetInfo] does not update the handle's State.
+func (h *OperationHandle) State() nexusapi.OperationState {
+	return h.state
+}
+
+// Close the handle's associated http response, if present.
+func (h *OperationHandle) Close() error {
+	// Body will have already been closed
+	if h.state != nexusapi.OperationStateSucceeded {
+		return nil
+	}
+	return h.response.Body.Close()
+}
+
+// Options for [nexusclient.OperationHandle.GetInfo].
+type GetInfoOptions struct {
+	// Header to attach to the HTTP request. Optional.
+	Header http.Header
+}
+
+// GetHandle gets a handle to an asynchronous operation by name and ID.
+// Does not incur a trip to the server.
+func (c *Client) GetHandle(operation string, operationID string) *OperationHandle {
+	return &OperationHandle{
+		client:    c,
+		operation: operation,
+		id:        operationID,
+	}
+}
+
+// GetInfo gets operation information issuing a network request to the service handler.
+//
+// ⚠️ Does not update the handle's State.
+func (h *OperationHandle) GetInfo(ctx context.Context, options GetInfoOptions) (*nexusapi.OperationInfo, error) {
+	if h.id == "" {
+		return nil, ErrHandleForSyncOperation
+	}
+	return h.client.getOperationInfo(ctx, getOperationInfoRequest{
+		Operation:   h.operation,
+		OperationID: h.id,
+		Header:      options.Header,
+	})
+}
+
+type getOperationInfoRequest struct {
+	Operation   string
+	OperationID string
+	Header      http.Header
+}
+
+func (c *Client) getOperationInfo(ctx context.Context, request getOperationInfoRequest) (*nexusapi.OperationInfo, error) {
+	url, err := c.joinURL(request.Operation, request.OperationID)
+	if err != nil {
+		return nil, err
+	}
+	httpReq, err := http.NewRequestWithContext(ctx, "GET", url.String(), nil)
+	if err != nil {
+		return nil, err
+	}
+	if request.Header != nil {
+		httpReq.Header = request.Header.Clone()
+	}
+
+	httpReq.Header.Set(headerUserAgent, UserAgent)
+	response, err := c.Options.HTTPClient.Do(httpReq)
+	if err != nil {
+		return nil, err
+	}
+	defer response.Body.Close()
+
+	if response.StatusCode != 200 {
+		return nil, c.newUnexpectedResponseError(fmt.Sprintf("unexpected response status: %q", response.Status), response)
+	}
+
+	return c.operationInfoFromResponse(response)
+}
+
+// Options for [nexusclient.OperationHandle.GetResult].
+type GetResultOptions struct {
+	// Header to attach to the HTTP request. Optional.
+	Header http.Header
+	// Boolean indicating whether to wait for operation completion or return the current status immediately.
+	Wait bool
```

Comment:
I'd rather avoid the ambiguity but I do get your point that to wait for a shorter duration than the deadline, users will need to construct a second context and set the deadline.
Let me consider this.

Reply:
I think reusing context deadline for both the HTTP call _and_ the wait time is causing ambiguity. If you want to avoid the ambiguity, take my suggestion of separating them.